### PR TITLE
add hostdir e2e test

### PR DESCRIPTION
**Pull request by dalanlan**
test passed on my own repo dalanlan/host-dir-mounttest

Signed-off-by: He Simei hesimei@zju.edu.cn

### CODE REVIEW

File: contrib/for-tests/hostdir-mount-tester/mt-hostdir.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+var (
+	fileModePath        = ""
+	readFileContentPath = ""
+	writeNewFilePath    = ""
+)
+
+func init() {
+	flag.StringVar(&fileModePath, "file_mode", "", "Path to print the filemode of")
```
Comment (roberthbailey) [inline contrib/for-tests/hostdir-mount-tester/mt-hostdir.go]: Why not use flag.String(...) above where the variables are declared?

**Comment by pmorie** [comment]
Doesn't this change touch secrets e2e too?

File: contrib/for-tests/hostdir-mount-tester/mt-hostdir.go (hunk 1 above)
Comment (dalanlan) [inline contrib/for-tests/hostdir-mount-tester/mt-hostdir.go]: Just a habit.
See also
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cmd/e2e/e2e.go#L46
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/contrib/for-tests/mount-tester/mt.go#L35

**Comment by dalanlan** [comment]
I could use some advice.
When i conducted local test, i would run `make` manually under the contrib/for-tests/mount-host-dir-tester to generate available binary and image, where occurred as a problem here. 
Any help would be appreciated.

**Comment by thockin** [comment]
I really appreciate getting this test!  I wonder if there is an easier attack, though?

Would it be sufficient to simply schedule a pod onto a known node, where the pod is simply a busybox script that writes the hostname to a fixed path and sleeps forever?  verification becomes simply ssh to node, ensure that the correct file was written on the host,


container: busybox sh -c "hostname > /host/hostname; sleep 3600"
verify: ssh to node & read /tmp/host-dir-volume-test


Maybe I am nuts - I just get worried that we have a growing number of containers floating around that we have to build and push and pull...

**Comment by dalanlan** [comment]
Would it make any sense to you if i just use my own repo dalanlan/host-dir-mounttest as the image of the containers? Then we could save the build and push work, sacrificing transparency may harm though. @thockin 
Or i'll take your advice with pleasure:)

**Comment by pmorie** [comment]
The mount tester image can write a file for you.

**Comment by thockin** [comment]
We don't want to rely on images we can't control and change as needed, so
ideally we use either a very generic one like busybox or one we control
like you have proposed.  I'm OK with either, I guess.

**Comment by dalanlan** [comment]
I use the existing image kubernetes/mounttest:0.1 instead. I doubt whether it makes sense at the beginning since it was built from scratch.
Travis-cli passed. plz kindly have a check:)
@pmorie @thockin

File: test/e2e/host_dir.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		c         *client.Client
+		podClient client.PodInterface
+	)
+
+	BeforeEach(func() {
+		var err error
+		c, err = loadClient()
+		expectNoError(err)
+
+		podClient = c.Pods(api.NamespaceDefault)
```
Comment (quinton-hoole) [inline test/e2e/host_dir.go]: Please avoid using the default namespace for e2e tests.  Please use a unique namespace for each invocation of each e2e test.  That way we can run multiple e2e's in parallell, and also, if they don't clean up properly after themselves, they will not cause subsequent runs to fail.  
You can use createTestingNamespace() for this. See https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/networking.go for an example.

File: test/e2e/host_dir.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	})
+
+})
+
+const containerName1 = "test-container-1"
+const containerName2 = "test-container-2"
+const hostdirvolumeName = "test-volume"
+
+func testPodWithHostVolume(path string, source *api.HostPathVolumeSource) *api.Pod {
+	podName := "pod-" + string(util.NewUUID())
```
Comment (quinton-hoole) [inline test/e2e/host_dir.go]: Given the above comment regarding running in a unique namespace, you could use a simple literal name here.

**Comment by dalanlan** [comment]
Fixed:)
@quinton-hoole

File: contrib/for-tests/mount-tester/mt.go [hunk 4]
```diff
@@ -143,3 +151,30 @@ func readWriteNewFile(path string) error {
...
+func readFileContentInLoop(path string) error {
+	if path == "" {
+		return nil
+	}
+	var (
+		contentBytes []byte
+		err          error
+	)
+	start := time.Now()
+	for time.Now().Sub(start) < (500 * time.Second) {
```
Comment (quinton-hoole) [inline contrib/for-tests/mount-tester/mt.go]: Please rather use our standard retry library in wait.Poll()

File: contrib/for-tests/mount-tester/mt.go [hunk 5]
```diff
@@ -143,3 +151,30 @@ func readWriteNewFile(path string) error {
 
 	return readFileContent(path)
 }
+
+func readFileContentInLoop(path string) error {
```
Comment (quinton-hoole) [inline contrib/for-tests/mount-tester/mt.go]: This function only checks that >0 bytes can be read from the file, and is not configurable with respect to how many bytes are expected, how long to retry for etc.

Would it not make sense to have it take bytesExpected and retryDuration as input parameters to make it more generally useful?  And then perhaps move it to one of the utils packages (e.g. testing/utils)?

**Comment by quinton-hoole** [comment]
Other than the two minor comment, LGTM. 
My apologies for not spotting those in my first review.

File: contrib/for-tests/mount-tester/mt.go (hunk 4 above)
Comment (quinton-hoole) [inline contrib/for-tests/mount-tester/mt.go]: Please use our standard wait.Poll() utility function here, rather than this loop.

**Comment by quinton-hoole** [comment]
@dalanlan Please let me know when you've addressed the remaining feedback.  Thanks.

**Comment by dalanlan** [comment]
I did that part the day before yesterday, but travis-ci seemed to hang there.
I'll see what I can do. @quinton-hoole

**Comment by dalanlan** [comment]
WHAT A SHAME!
`e2e/host_path.go` has already been there. You could simply close this pr:(
@quinton-hoole

**Comment by dalanlan** [comment]
Ahhhh, i had a problem with `host_path.go` here.
Is it gorgeous enough to test write & read with a single container? I thought the writing test should be checked by another container so that we can make sure the volume is shared between pod.
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/host_path.go#L90

**Comment by quinton-hoole** [comment]
@pmorie I don't know much about secrets.  What did you have in mind in your comment above.

cc: @erictune FYI

@dalanlan Yes, I think that we should test writing and reading from different containers. I hope to complete this review soon :-)

File: test/e2e/util.go [hunk 6]
```diff
@@ -936,3 +941,30 @@ func HighLatencyRequests(c *client.Client, threshold time.Duration, ignoredResou
...
 	return len(badMetrics), nil
 }
+
+func TestFileContent(filePath string, bytesExpected int, retryDuration int) ([]byte, error) {
+	var (
+		contentBytes []byte
+		err          error
+	)
+
+	expectNoError(wait.Poll(time.Second*2, time.Second*time.Duration(retryDuration), func() (bool, error) {
```
Comment (dalanlan) [inline test/e2e/util.go]: Need to use `for` loop here instead, right?

Comment (quinton-hoole) [inline test/e2e/util.go]: No, wait.Poll does the loop for you.

**Comment by erictune** [comment]
status: awaiting review.

**Comment by quinton-hoole** [comment]
It's not clear whether @dalanlan still want's this PR merged?  Her comment on May 17, 5:54pm ("OH WHAT A SHAME... You could simply close this pr:("  indicates perhaps not.  Closing.  Please re-open if you still require it to be reviewed and merged.

**Comment by dalanlan** [comment]
Oops...
i still want it to be reviewed (volumes shared among different containers would be better), yet i guess it would be better to open another pr to finish this job for it may cost a lot to rebase:) 
Will do it later.

**Comment by dalanlan** [comment]
@quinton-hoole

**Pull request closed without merging**